client/request_test.go

```diff
@@ -0,0 +1,72 @@
+package client
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"testing"
+
+	"golang.org/x/net/context"
+)
+
+// TestSetHostHeader should set fake host for local communications, set real host
+// for normal communications.
+func TestSetHostHeader(t *testing.T) {
+	testURL := "/test"
+	testCases := []struct {
+		host            string
+		expectedHost    string
+		expectedURLHost string
+	}{
+		{
+			"unix:///var/run/docker.sock",
+			"",
+			"docker",
+		},
+		{
+			"npipe:////./pipe/docker_engine",
+			"",
+			"docker",
+		},
+		{
+			"tcp://0.0.0.0:4243",
+			"",
```

Comment:
OK. If we are using something like "tcp://0.0.0.0:4243//abc/efg", the `request.Host` will be set to `"abc"`.

Because the `basePath` will become "//abc/efg", then the `apiPath` will become something like "//abc/efg/container/id/start", then `http.NewRequest` will set `request.Host` to "abc".

I think this is also a bug in engine-api, we should fix it instead of testing it.